### ISSUE TITLE
Is it possible to run two windows in two threads?

### ISSUE THREAD

**Issue by MaldoPay** [raysan5/raylib#2940]
I have this:

```
    void start_window(std::string windowName){
    
        InitWindow(800, 450, windowName.c_str());
        SetTargetFPS(60);
        rlFPCamera cam;
        cam.Setup(45,Vector3{ 1, 0, 0 });
        cam.SetCameraPosition((Vector3){1.0f, 500.0f, 1.0f});
        cam.MoveSpeed.z = 100;
        cam.MoveSpeed.x = 100;
        cam.FarPlane = 5000;
    
    
        Material matDefault = LoadMaterialDefault();
        Material matRed = LoadMaterialDefault();
        Material matGreen = LoadMaterialDefault();
        matDefault.maps[MATERIAL_MAP_DIFFUSE].color = BLUE;
        matRed.maps[MATERIAL_MAP_DIFFUSE].color = RED;
        matGreen.maps[MATERIAL_MAP_DIFFUSE].color = GREEN;
    
        Model cube = LoadModelFromMesh(GenMeshCube(5,5,5));
        int width = 3072;
        int height = 3072;
    
        // Dynamic memory allocation to store pixels data (Color type)
        Color *pixels = (Color *)malloc(width*height*sizeof(Color));
    
        for (int y = 0; y < height; y++)
        {
            for (int x = 0; x < width; x++)
            {
                if (((x/16+y/16)/1)%2 == 0) pixels[y*width + x] = WHITE;
                else pixels[y*width + x] = BLUE;
            }
        }
        Image checkedIm = {
                .data = pixels,             // We can assign pixels directly to data
                .width = width,
                .height = height,
                .mipmaps = 1,
                .format = PIXELFORMAT_UNCOMPRESSED_R8G8B8A8,
        };
    
        Texture2D checked = LoadTextureFromImage(checkedIm);
    
        Model terrain = LoadModel("Resources/Terrains/TestingGround.obj");
        terrain.materials[0].maps[MATERIAL_MAP_DIFFUSE].texture = checked;
        Vector3 cubePosition = {2.50f, 3.0f, 2.50f};
    
        UnloadImage(checkedIm);
        while (!WindowShouldClose())    // Detect window close button or ESC key
        {
            cam.Update();
    
            float offsetThisFrame = 100.0f*GetFrameTime();
    
    
            BeginDrawing();
            ClearBackground(SKYBLUE);
            cam.BeginMode3D();
            DrawGrid(1000, 1.0f);
            if (IsKeyDown(KEY_SPACE))
            {
                cam.SetCameraPosition((Vector3){cam.GetCameraPosition().x, cam.GetCameraPosition().y+offsetThisFrame, cam.GetCameraPosition().z});
            }
            else if (IsKeyDown(KEY_Z))
            {
                cam.SetCameraPosition((Vector3){cam.GetCameraPosition().x, cam.GetCameraPosition().y-offsetThisFrame, cam.GetCameraPosition().z});
            }
            if (IsKeyPressed(KEY_SPACE))
            {
                cubePosition = {cubePosition.x, cubePosition.y + 5, cubePosition.z};
            }
            DrawModel(cube, (Vector3){50.30f, 12.32f, 40.7f}, 1, RED);
            DrawModel(cube, (Vector3){68.8f, 8.8f, 74.7f}, 1, RED);
            DrawModel(cube, (Vector3){15.52f, 11.65f, 84.51f}, 1, RED);
            DrawModel(cube, cubePosition, 1, BLUE);
    
            DrawModelEx(terrain, (Vector3){0.0f, 0.0f, 0.0f},(Vector3){0.0f, 0.0f, 0.0f}, 0.0f,(Vector3){1.0f, 1.0f, 1.0f}, WHITE);
    
            EndMode3D();
    
            DrawFPS( 10, 10);
            DrawText("Press H to hide the Map Window.", 10, 32, 12, GREEN);
            DrawText(TextFormat("Camera Pos: X: %3.2fY:  %3.2f Z: %3.2f",
                                cam.GetCameraPosition().x * -1,
                                cam.GetCameraPosition().y,
                                cam.GetCameraPosition().z), 10, 52, 12, BLACK);
            EndDrawing();
        }
        UnloadModel(cube);
        CloseWindow();
    }
```
Than I do the following:

```
    std::thread game1 (&start_window, "test");
    std::thread game2 (&start_window, "test 2");
```

Looks like it brakes when I add the second thread. Any idea how can I fix that? Is it even possible to run two windows in two different threads ?

**Comment by MaldoPay** [comment]
If running two windows from two threads is not possible, can you somehow at least run more than 1 window and render more than 1 scene at a time ?

**Comment by raysan5** [comment]
@MaldoPay Multiple windows is out of scope for raylib, you can just implement it for your use-case following this guide: https://github.com/raysan5/raylib/wiki/Use-multiple-windows

**Comment by MaldoPay** [comment]
@raysan5 thank you! Does that mean the two windows will run in two separate threads that way ?

**Comment by raysan5** [comment]
@MaldoPay I think so but OpenGL is single-thread, still you can have two separate OpenGL context... you should try.

**Comment by MaldoPay** [comment]
@raysan5 I tried the guide you shared but I think it's outdated maybe. When I searched for `MAX_CONTEXTS ` in raylib.h as it is described in the guide. There is no such thing. Are you sure this guide is still valid ?

**Comment by raysan5** [comment]
@MaldoPay raylib does not support multiple windows, you have to modify raylib yourseld to add support for it if you need it. The guide I shared is the modifications required.

**Comment by MaldoPay** [comment]
@raysan5 I did all my best to bring the guide to the actual version of RayLib. The link to repositories showed really old version of RayLib like 3.0. 

So I created a fork in order to give it a go. If you take a quick look: https://github.com/MaldoPay/raylib/commit/db9018836846b8e3b0b6bdb18b6b1ecedc0d4efc

And maybe compile it you can see that it crashes on line **1144**:

```
1141	// Reset current matrix to identity matrix
1142	void rlLoadIdentity(void)
1143	{
1144	    *RLGL[currentContext].State.currentMatrix = rlMatrixIdentity();
1145	}
```

Can you please take a look and give any suggestion/solution on how can I achieve multiple windows with most recent version of RayLib.

P.S.
I am struggling on this for several days. So your help is a HUGE thing for me :)

**Comment by orcmid** [comment]
@MaldoPay I am not an expert about this, but I think the reason there can be no multiple windows in raylib is explained by `rcore.c` line 507 or thereabouts.

```C
static CoreData CORE = { 0 }
```

The struct type, `CoreData` is defined starting at lines 375-507.

That's the TL;DR.  Here's more thinking about this.

The key thing is there is only one of these as global data to all of the raylib functions in a single `.exe` and it is designed around there being only one window (although maybe multiple frames in that one window).

There are no separate instances and I have no idea what is done to have raylib shared libraries work properly when shared with two running raylib applications.  There are ways that sort of thing can be handled in the initialization of a shared library for use by more than one process, but I have no idea if that has been done for raylib.

This is sort of why wrappers/bindings in object-oriented models, such as for C++, are a bit sketchy.  The CORE becomes, in effect, data of a static object instance and not something that can have multiple instances (and solve your problem thereby).  There's a pile of things to have handled (pun intended), including mouse and keyboard focus, etc.

To get around this, and I think that is what Ramone is pointing at, you would need to build a separate version of raylib and maintain it in parallel with the official raylib releases.   Is it worth it to you?

I wonder if any pro game engines do this any better.  Have you seen a game or application that operates multiple windows?  The GUI system does, but that's a whole different thing with process and thread management and switching of attention.

**Comment by orcmid** [comment]
@MaldoPay: An afterthought.

Have you considered having two programs, where the main one initiates the other one and they communicate by some inter-application connection?  (I am thinking pipes or something otherwise allowing mutual synchronization and communication.)   That would allow for two different instances of raylib operation and two separate but coordinated windows.

This could also be terribly gnarly and a worse problem than whatever it is you are trying to solve by operating two windows.  The description of changes by @raysan5 is probably challenging enough.

**Comment by MaldoPay** [comment]
@orcmid This sounds like really bad design. What I want to achieve is to use raylib as a "game engine" solution for C++ server solution. However @raysan5 can confirm but I think raylib have no intention in adding features for multiple window/scenes solutions?

P.S.
@raysan5 did you had chance to look at the fork I've made?

**Comment by raysan5** [comment]
@MaldoPay Multiple-window support is out-of-scope for raylib at the moment. Sorry, I'm not looking into it, maybe in a future.

**Comment by orcmid** [comment]
@MaldoPay I agree that would be a bad design, especially for what you want.  If you are looking to have a (multi-player?) game in the cloud, I think professional game engines are required.  There's much around doing such a thing using Azure; Microsoft provides encouraging resources. 

I suspect there is equivalent support for games to be released on PlayStation, Xbox, and other consoles, as well as ones working with web browser and smartphone clients. 

If your needs are more special-purpose than that, you might check around with other graphic-support projects.
